src/handlers/applicationFormsHandlers.ts

```diff
@@ -0,0 +1,125 @@
+import type {
+  Request,
+  Response,
+} from 'express';
+import type { Result } from 'tinypg';
+import type { JSONSchemaType } from 'ajv';
+import type { ApplicationForm } from '../types';
+
+const logger = getLogger(__filename);
+
+const getApplicationForms = (req: Request, res: Response): void => {
+  db.sql('applicationForms.selectAll')
+    .then((applicationFormsQueryResult: Result<ApplicationForm>) => {
+      logger.debug(applicationFormsQueryResult);
+      const { rows: applicationForms } = applicationFormsQueryResult;
+      if (isApplicationFormArray(applicationForms)) {
+        res.status(200)
+          .contentType('application/json')
+          .send(applicationForms);
+      } else {
+        throw new ValidationError(
+          'The database responded with an unexpected format.',
+          isApplicationFormArray.errors ?? [],
+        );
+      }
+    })
+    .catch((error: unknown) => {
+      logger.error(error);
+      res.status(500)
+        .contentType('application/json')
+        .send(error);
+    });
+};
+
+const postApplicationFormsBodySchema: JSONSchemaType<Omit<ApplicationForm, 'createdAt' | 'id' | 'version'>> = {
+  type: 'object',
+  properties: {
+    opportunityId: {
+      type: 'number',
+    },
+  },
+  required: [
+    'opportunityId',
+  ],
+};
+const isPostApplicationFormsBody = ajv.compile(postApplicationFormsBodySchema);
+const postApplicationForms = (
+  req: Request<unknown, unknown, Omit<ApplicationForm, 'createdAt' | 'id' | 'version'>>,
+  res: Response,
+): void => {
+  if (!isPostApplicationFormsBody(req.body)) {
+    res.status(400)
+      .contentType('application/json')
+      .send({
+        message: 'Invalid request body.',
+        errors: isPostApplicationFormsBody.errors,
+      });
+    return;
+  }
+
+  db.sql('applicationForms.insertOne', {
+    ...req.body,
+    optedIn: false,
+  })
+    .then((applicationFormsQueryResult: Result<ApplicationForm>) => {
+      logger.debug(applicationFormsQueryResult);
+      const applicationForm = applicationFormsQueryResult.rows[0];
+      if (isApplicationForm(applicationForm)) {
+        res.status(201)
+          .contentType('application/json')
+          .send(applicationForm);
+      } else {
+        throw new ValidationError(
+          'The database responded with an unexpected format.',
+          isApplicationForm.errors ?? [],
+        );
+      }
+    })
+    .catch((error: unknown) => {
+      if (error instanceof TinyPgError
+      && isQueryContextWithError(error.queryContext)) {
+        switch (error.queryContext.error.code) {
+          case PostgresErrorCode.FOREIGN_KEY_VIOLATION:
+            res.status(409)
+              .contentType('application/json')
+              .send({
+                message: 'Unique key constraint violation.',
+                errors: [error.queryContext.error],
+              });
+            return;
+          case PostgresErrorCode.UNIQUE_VIOLATION:
```

Comment:
Actually, since the unique constraint is based on `version` and we don't let the user specify a version / that is maintained on our end, I don't believe a unique validation *can* occur (and if it does it is a 5xx error not a 4xx error).  given that, I don't think it's possible to write a test that generates this case and so I'm gonna just take it out and if somehow that comes through it would go down to the 500 error logic below.

Reply:
I've moved this to a simple if statement to cover that single known likely; in a future PR I'm going to propose an alternative way to handle these errors anyway so that it's always robust regardless of what we understand as possible in a single case..